src/Server/TCPHandler.cpp

```diff
@@ -1212,13 +1212,23 @@ void TCPHandler::receiveQuery()
             throw NetException("Hash mismatch", ErrorCodes::UNEXPECTED_PACKET_FROM_CLIENT);
         /// TODO: change error code?
 
-        /// initial_user can be empty in case of Distributed INSERT via Buffer/Kafka,
-        /// i.e. when the INSERT is done with the global context (w/o user).
-        if (!client_info.initial_user.empty())
+        /// Previously, initial_user can be empty in case of Distributed INSERT via Buffer/Kafka,
+        /// (i.e. when the INSERT is done with the global context w/o user),
+        /// but this will make client to use previously authenticated user,
+        /// which is bad (it may have different settings, and even readonly=1)
+        ///
+        /// Also note, that we want to change client_info, to reflect the change in query_log.
+        std::string & user = client_info.initial_user;
+        if (user.empty())
         {
-            LOG_DEBUG(log, "User (initial): {}", client_info.initial_user);
-            session->authenticate(AlwaysAllowCredentials{client_info.initial_user}, client_info.initial_address);
+            user = "default";
```

Comment:
For example `Buffer` inserts into its underlying table using the global context, and the global context isn't always the same as the context of the user named `default`. So maybe in this case when we get a secondary query here with an empty initial user we should assume it's the global context.

And assuming that if `user.empty()` we should call here something like
`session->resetAuthentication()` instead of calling
`session->authenticate(AlwaysAllowCredentials{"default"})`.

Reply:
>For example Buffer inserts into its underlying table using the global context, and the global context isn't always the same as the context of the user named default. 

Agree (that's why I wrote "at least" in the commit message :) )

> So maybe in this case when we get a secondary query here with an empty initial user we should assume it's the global context.

Indeed.

> And assuming that if user.empty() we should call here something like
session->resetAuthentication() instead of calling
session->authenticate(AlwaysAllowCredentials{"default"}).

Yep, great idea - done.